src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -438,6 +449,47 @@ private void downloadFileIfRemote(PathFragment path) throws IOException {
     }
   }
 
+  @Override
+  public void renameTo(PathFragment sourcePath, PathFragment targetPath) throws IOException {
+    FileNotFoundException remoteException = null;
+    try {
+      remoteOutputTree.renameTo(sourcePath, targetPath);
+    } catch (FileNotFoundException e) {
+      remoteException = e;
+    }
+
+    FileNotFoundException localException = null;
+    try {
+      delegateFs.renameTo(sourcePath, targetPath);
+    } catch (FileNotFoundException e) {
+      localException = e;
+    }
+
+    if (remoteException == null || localException == null) {
+      return;
+    }
+
+    localException.addSuppressed(remoteException);
+    throw localException;
+  }
+
+  @Override
+  public void createDirectoryAndParents(PathFragment path) throws IOException {
+    super.createDirectoryAndParents(path);
+    if (path.startsWith(outputBase)) {
+      remoteOutputTree.createDirectoryAndParents(path);
+    }
+  }
+
+  @Override
+  public boolean createDirectory(PathFragment path) throws IOException {
+    boolean created = delegateFs.createDirectory(path);
```

Comment:
Same as above - is it stricly necessary to create the directory in the local filesystem?

Reply:
yes, for dynamic execution AND for action logs.